libsolidity/codegen/ContractCompiler.cpp

```diff
@@ -1361,11 +1361,20 @@ void ContractCompiler::appendModifierOrFunctionCode()
 	m_context.setModifierDepth(m_modifierDepth);
 }
 
-void ContractCompiler::appendStackVariableInitialisation(VariableDeclaration const& _variable)
+void ContractCompiler::appendStackVariableInitialisation(
+	VariableDeclaration const& _variable,
+	bool _provideDefaultValue
+)
 {
 	CompilerContext::LocationSetter location(m_context, _variable);
 	m_context.addVariable(_variable);
-	CompilerUtils(m_context).pushZeroValue(*_variable.annotation().type);
+	if (!_provideDefaultValue && _variable.type()->dataStoredIn(DataLocation::Memory))
+	{
+		solAssert(_variable.type()->sizeOnStack() == 1, "");
+		m_context << u256(0);
```

Comment:
`pushZeroValue()` internally uses `pushZeroPointer()` for dynamically-sized arrays, which currently is `0x60`, not `0`.  This overrides that code. Is that on purpose?

Could you add a test for a case where the difference actually matters?

Reply:
It matters for structs and non-dynamically-sized arrays. See the test you already commented on.

Reply:
But I'm asking about dynamically sized arrays because I think this condition will catch all memory types, including both statically- and dynamically-sized arrays.

> It matters for structs and non-dynamically-sized arrays. See the test you already commented on.

I think it actually **does not** matter for structs and non-dynamically-sized arrays, because anything we put in it will get overwritten by the initial value, right?

Reply:
What I meant was saving memory is possible for structs and statically-sized arrays.

Reply:
So I'm saying that this change seems to affect more than just structs and statically-sized arrays. Won't e.g. `uint[] memory x;` satisfy `!_provideDefaultValue && _variable.type()->dataStoredIn(DataLocation::Memory)` too?

Will the behavior of the code remain the same? If so, why?
```solidity
function f() {
    uint[] memory x;
    uint y;
    assembly {
        y := x
    }
    assert(y == 0);
}
```

Reply:
It should be the same because `x` does not get assigned a value right away. Maybe to clarify: Before this change, the line `StructName memory x = y;` would allocate a fresh zero-initialized struct for x, immediately forget about it again and assign a value to `x` that points to the same data as `y`.

Reply:
Okay, test passed so I guess it's fine but honestly, I still don't understand why. For dynamic arrays this code does `m_context << u256(0)` while `pushZeroValue()` was doing `m_context << u256(zeroPointer)`. Are you saying that even in case of `uint[] memory x;`, the zero in `x` just gets overwritten with `zeroPointer` later (even if there's no explicit initialization)?